src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs

```diff
@@ -290,6 +290,47 @@ internal partial class Binder
                             syntaxBuilder!.Add(typeConstraintSyntax);
                         }
                         continue;
+
+                    case SyntaxKind.AllowsConstraintClause:
+
+                        if (isForOverride)
+                        {
+                            reportOverrideWithConstraints(ref reportedOverrideWithConstraints, syntax, diagnostics);
+                            continue;
+                        }
+
+                        if (i != n - 1)
+                        {
+                            diagnostics.Add(ErrorCode.ERR_AllowsClauseMustBeLast, syntax.GetFirstToken().GetLocation());
+                        }
+
+                        bool hasRefStructConstraint = false;
+
+                        foreach (var allowsConstraint in ((AllowsConstraintClauseSyntax)syntax).Constraints)
+                        {
+                            if (allowsConstraint.Kind() == SyntaxKind.RefStructConstraint)
+                            {
+                                if (hasRefStructConstraint)
+                                {
+                                    diagnostics.Add(ErrorCode.ERR_RefStructConstraintAlreadySpecified, allowsConstraint);
+                                }
+                                else
+                                {
+                                    CheckFeatureAvailability(allowsConstraint, MessageID.IDS_FeatureRefStructInterfaces, diagnostics);
+
+                                    if (!Compilation.Assembly.RuntimeSupportsByRefLikeGenerics)
+                                    {
+                                        Error(diagnostics, ErrorCode.ERR_RuntimeDoesNotSupportByRefLikeGenerics, allowsConstraint);
+                                    }
+
+                                    constraints |= TypeParameterConstraintKind.AllowByRefLike;
+                                    hasRefStructConstraint = true;
+                                }
+                            }
+                        }
+
+                        continue;
```

Comment:
> Can we get here if ```hasRefStructConstraint``` is `false` and if so, should we report an error?

I do not think it is possible to get such a tree from the parser. The syntax node description requires at least one item in the list. So, if someone somehow manages to create an `allows` clause with an empty list of constraints, that is on them and I am very comfortable to not report any errors about that. We generally do not guard against shapes that cannot come from our parser.

Reply:
> We probably don't want to have an assert here either, since one could create the tree manually, right?

Right. The code is not making an assumption that the list is not empty. Therefore, having the assert is probably inappropriate.